# Episodic Memory with Write-Time Reflection

This notebook demonstrates the `reflective_memory()` factory, which gives an
`LLMAgent` episodic memory that distils a short lesson from each completed task at
write time.

The example is drawn directly from the
[Reflexion paper (Shinn et al., 2023)](https://arxiv.org/abs/2303.11366), which
frames the pattern as verbal reinforcement learning: a self-reflection model converts
a completed episode into a natural-language lesson, and episodic memory stores that
lesson for future retrieval.

The task is Monte Carlo estimation of pi — one round, two tool calls. The agent calls
`generate_random_sample(n)` to seed a sample and `monte_carlo_estimate` to get the
estimate; the only decision is choosing `n`. A custom `template` supplies the true
value of pi, giving the reflection model a concrete error signal: it can reason about
how far off the estimate was and recommend a specific sample size for next time.

Two properties are illustrated:

- **Write-time reflection.** After the task completes, `reflective_memory()` calls
  the LLM to produce a one-sentence lesson from the task and its outcome. That lesson
  is attached to the episode under `metadata["reflection"]` and rendered automatically
  inside a `<reflection>` tag whenever the episode is displayed.
- **Recalled lessons.** When a new task arrives, `recall()` surfaces the most
  semantically similar episode. Because the episode carries a pre-distilled lesson,
  the agent can apply the insight on the next attempt rather than rediscovering the
  same mistakes.

<!-- provider-note:start -->
## Setting the backbone LLM of your agent

These notebooks run on **Ollama by default**, the setup the book teaches.
If you do nothing, nothing changes: `make_llm()` starts a local Ollama
service when one isn't already running.

To use OpenAI or Anthropic instead:

1. Install the extra: `uv sync --extra openai` or `uv sync --extra anthropic`.
2. Export `OPENAI_API_KEY` or `ANTHROPIC_API_KEY` before launching Jupyter.
3. Set `LLM_PROVIDER=openai` (or `anthropic`), or pass `provider="openai"`
   to `make_llm()`. A key on its own never switches providers, so one
   exported for unrelated work cannot reroute you off the Ollama path.

The switch applies wherever a notebook builds its LLM with `make_llm()`.
A notebook that constructs `OllamaLLM` directly stays on Ollama regardless
of these settings.

If you opted in but forgot to export the key, you will be prompted for it.
That is the safe path on hosted kernels, and it keeps the key out of the
saved notebook. Setting `OLLAMA_API_KEY` alone routes Ollama to Ollama
Cloud.

**Caveat:** the examples are tuned for qwen3. Output on gpt-5 or Claude
will differ from what is printed in the book, and prompt-sensitive
examples (the ch09 evaluator pattern, ch08 supervised trajectories) may
behave noticeably differently.
<!-- provider-note:end -->

In [1]:
# Uncomment the line below to install `llm-agents-from-scratch` from PyPI
# !pip install llm-agents-from-scratch

## Running an Ollama service

To execute the code in this notebook you need Ollama installed locally with its
service running. Download Ollama from https://ollama.com/download, then start the
service by running `ollama serve` in a terminal.

## Setup

In [2]:
from llm_agents_from_scratch.notebook_utils import make_llm

In [3]:
import logging
import tempfile
from pathlib import Path

from llm_agents_from_scratch import LLMAgent
from llm_agents_from_scratch.data_structures import Task
from llm_agents_from_scratch.logger import enable_console_logging
from llm_agents_from_scratch.memory.recipes import (
    recency_memory,
    reflective_memory,
)

## Part 1 — Tools for Monte Carlo Pi Estimation

The tools here are adapted from Capstone 1, which applies the same Monte Carlo
algorithm in a more open-ended multi-step task. Two of the three capstone tools are
wired to the agent in this demo:

- `generate_random_sample(n)` — generates `n` random 2-D points in [0, 1] × [0, 1],
  stores them in a global registry, and returns a `sample_id`.
- `monte_carlo_estimate(sample_id)` — counts the fraction of stored points that fall
  inside the unit circle, multiplies by 4, and returns the estimate alongside the
  sample size.

Capstone 1 also defines a third tool, `add_more_points_to_sample`, which lets an
agent grow an existing sample incrementally without discarding accumulated data. It
is omitted here because the task constrains the agent to one round — a single
`generate_random_sample` call followed by a single `monte_carlo_estimate`. The only
decision is choosing `n`.

In [4]:
import uuid

import numpy as np
from pydantic import BaseModel, ConfigDict, Field, computed_field

from llm_agents_from_scratch.tools import PydanticFunctionTool

SAMPLE_REGISTRY: dict[str, list[tuple[float, float]]] = {}


class RandomSampleParams(BaseModel):
    """Params for generate_random_sample."""

    model_config = ConfigDict(extra="forbid")
    n: int = Field(description="Number of random points to generate.")


class RandomSample(BaseModel):
    """Result from generate_random_sample."""

    sample_id: str = Field(
        description="Pass this sample_id to monte_carlo_estimate.",
    )

    @computed_field
    @property
    def sample_size(self) -> int:
        """Current number of points in the sample."""
        return len(SAMPLE_REGISTRY[self.sample_id])

    def __str__(self) -> str:
        """JSON string representation."""
        return self.model_dump_json()


def generate_random_sample(params: RandomSampleParams) -> RandomSample:
    """Generate n random points in [0,1]x[0,1] and store them.

    Returns a sample_id. Pass this sample_id to monte_carlo_estimate.
    """
    pts = np.random.uniform(size=(params.n, 2))
    sample_id = str(uuid.uuid4())
    SAMPLE_REGISTRY[sample_id] = [tuple(pt) for pt in pts.tolist()]
    return RandomSample(sample_id=sample_id)


class MonteCarloParams(BaseModel):
    """Params for monte_carlo_estimate."""

    model_config = ConfigDict(extra="forbid")
    sample_id: str = Field(
        description="The sample_id returned by generate_random_sample.",
    )


class MonteCarloResult(BaseModel):
    """Result from monte_carlo_estimate."""

    sample_id: str
    sample_size: int
    estimate: float

    def __str__(self) -> str:
        """JSON string representation."""
        return self.model_dump_json()


def monte_carlo_estimate(params: MonteCarloParams) -> MonteCarloResult:
    """Estimate pi using the Monte Carlo method.

    Counts the fraction of stored points that fall inside the unit circle
    and multiplies by 4.
    """
    points = SAMPLE_REGISTRY[params.sample_id]
    n = len(points)
    inside = sum(x**2 + y**2 < 1 for x, y in points)
    return MonteCarloResult(
        sample_id=params.sample_id,
        sample_size=n,
        estimate=(inside / n) * 4,
    )


random_sample_tool = PydanticFunctionTool(generate_random_sample)
mc_estimate_tool = PydanticFunctionTool(monte_carlo_estimate)

## Creating the Agent with Memory

`reflective_memory()` creates a `Memory` backed by an in-process Qdrant vector store.
Before each episode is written it calls `llm.complete()` to distil a one-sentence
lesson from the task and result, storing it under `metadata["reflection"]`.

The optional `template` parameter replaces the built-in reflection prompt. For this
demo it is replaced with a task-specific prompt that supplies the true value of pi.
This gives the reflection model an error signal: it can compute how far off the
estimate was and recommend a concrete sample size for the next attempt — rather than
giving generic advice about "using large samples".

For tasks where correctness is not obvious from the result alone, pass a custom
`template` that supplies a ground truth or evaluation rubric. The default template
works well when success and failure are unambiguous; a custom template is the hook
for turning a generic observation into a precise, actionable lesson.

The agent also carries a `recency_memory()` for the first run. It records the
same episode into a JSONL file in a temporary directory, with no reflection, so
the recency baseline below shows what a recency strategy would recall. It is
detached before the second run, so only the reflective memory guides it.

In [5]:
REFLECTION_TEMPLATE = """\
You just attempted to estimate pi using the Monte Carlo method.

Task: {instruction}
Result: {result}

The true value of pi is 3.141592653589793.

In one sentence, what specific sample size should be used next time to \
achieve a more accurate estimate, and why?\
"""

llm = make_llm()
memory = reflective_memory(llm=llm, max_results=1, template=REFLECTION_TEMPLATE)
recency = recency_memory(path=Path(tempfile.mkdtemp()), max_results=1)
agent = LLMAgent(
    llm=llm,
    tools=[random_sample_tool, mc_estimate_tool],
    memories=[memory, recency],
)

✓ Using Ollama Cloud (kimi-k2.7-code:cloud)


## Part 2 — Running the Pi Estimation Task

The agent gets one round: call `generate_random_sample(n)` to generate a sample,
then call `monte_carlo_estimate` to get the estimate. The only decision is choosing
`n`.

That single choice is where everything happens. Too small and the estimate is
unreliable; large enough and it converges to the right answer. The ALL CAPS warning
in the instruction is a hint — the reflection will capture whether the agent heeded
it.

This first run also caps the sample at one million points, so the estimate starts
with a real error for the reflection to learn from. The second run in Part 4 drops
the cap.

In [6]:
TASK_INSTRUCTION = (
    "Use the Monte Carlo method to estimate pi in one round.\n\n"
    "IMPORTANT: MONTE CARLO ESTIMATES OF PI CONVERGE VERY SLOWLY. "
    "YOU WILL NEED MILLIONS OF POINTS FOR A RELIABLE RESULT.\n\n"
    "Steps:\n"
    "1. Call generate_random_sample(n) — choose n to maximise accuracy.\n"
    "2. Call monte_carlo_estimate(sample_id) — this is your final estimate.\n\n"
    "You have two tools:\n"
    "- generate_random_sample(n): generates n random 2-D points in "
    "[0,1]x[0,1] and returns a sample_id.\n"
    "- monte_carlo_estimate(sample_id): computes a pi estimate from the "
    "stored sample.\n\n"
    "When done, respond with exactly:\n"
    '{"sample_id": "<the-sample-id>"}'
)

FIRST_RUN_INSTRUCTION = (
    TASK_INSTRUCTION
    + "\n\nFor this first attempt, use at most 1,000,000 points."
)

In [7]:
enable_console_logging(logging.INFO)

task1 = Task(instruction=FIRST_RUN_INSTRUCTION)
result1 = await agent.run(task1)
print(result1)

INFO (llm_agents_fs.LLMAgent) :      🚀 Starting task: Use the Monte Carlo method to estimate pi in one round.

IMPORTANT: MONTE CARLO ESTIMATES OF PI CONVERGE VERY SLOWLY. YOU WILL NEED M...[TRUNCATED]


INFO (llm_agents_fs.TaskHandler) :      ⚙️ Processing Step: Use the Monte Carlo method to estimate pi in one round.

IMPORTANT: MONTE CARLO ESTIMATES OF PI CONVERGE VERY SLOWLY. YOU WILL NEE...[TRUNCATED]
INFO (llm_agents_fs.TaskHandler) :      🛠️ Executing Tool Call: generate_random_sample
INFO (llm_agents_fs.TaskHandler) :      ✅ Successful Tool Call: {"sample_id":"8dbdb828-0e61-4280-81a8-002f4aad9e0a","sample_size":1000000}
INFO (llm_agents_fs.TaskHandler) :      ✅ Step Result: I need to make the following tool-calls:
{
    "id_": "e3a8b672-4a91-4064-aa1c-ae93944f31ff",
    "tool_name": "monte_carlo_estimate",
...[TRUNCATED]
INFO (llm_agents_fs.TaskHandler) :      🧠 New Step: Call the monte_carlo_estimate tool with sample_id "8dbdb828-0e61-4280-81a8-002f4aad9e0a" to produce the final pi estimate, then respond wi...[TRUNCATED]
INFO (llm_agents_fs.TaskHandler) :      ⚙️ Processing Step: Call the monte_carlo_estimate tool with sample_id "8dbdb828-0e61-4280-81a8-002f4aad9e0a" to produce

In [8]:
print(await memory.summary())

QdrantMemoryStore: 1 episodes | collection=episodes
  newest: 2026-09-30 01:39:12 | Use the Monte Carlo method to estimate pi in one round.

IMP
  oldest: 2026-09-30 01:39:12 | Use the Monte Carlo method to estimate pi in one round.

IMP
  metadata_fns: reflection


## The Recency Baseline

Before looking at the reflection, consider what `recency_memory()` would surface for
a new task. A recency-based memory stores no additional metadata. The recalled
episode carries the full result — which tools were called, what estimate was returned
— but the lesson is implicit.

A future agent seeing this raw episode would need to reason from the estimate alone
to figure out what went wrong and by how much. The signal is there, but buried.

The cell below reads it from the `recency_memory()` attached for the first run,
then detaches that memory so the second run sees only the reflective one.

In [9]:
print("What recency_memory() recalls (no reflection):\n")
for ep in await recency.store.recall(TASK_INSTRUCTION):
    print(str(ep))
    print()

agent.memories = [memory]

What recency_memory() recalls (no reflection):

  <episode>
    <task>Use the Monte Carlo method to estimate pi in one round.

IMPORTANT: MONTE CARLO ESTIMATES OF PI CONVERGE VERY SLOWLY. YOU WILL NEED MILLIONS OF POINTS FOR A RELIABLE RESULT.

Steps:
1. Call generate_random_sample(n) — choose n to maximise accuracy.
2. Call monte_carlo_estimate(sample_id) — this is your final estimate.

You have two tools:
- generate_random_sample(n): generates n random 2-D points in [0,1]x[0,1] and returns a sample_id.
- monte_carlo_estimate(sample_id): computes a pi estimate from the stored sample.

When done, respond with exactly:
{"sample_id": "<the-sample-id>"}

For this first attempt, use at most 1,000,000 points.</task>
    <result>The Monte Carlo estimate with 1,000,000 points is 3.139288.

{"sample_id": "8dbdb828-0e61-4280-81a8-002f4aad9e0a"}</result>
    <completed_at>2026-09-30 01:39:12</completed_at>
  </episode>



## Part 3 — Inspecting the Reflection

Now look at the same episode as stored by `reflective_memory()`. It carries a
`<reflection>` tag with the one-sentence lesson distilled at write time.

Because the custom `template` supplies the true value of pi, the reflection model
has a concrete error signal: it can compute how far off the estimate was and
recommend a specific sample size for next time. The lesson is captured in a single
sentence rather than buried in the full trajectory.

In [10]:
print("What reflective_memory() recalls (with reflection):\n")
for ep in await memory.store.recall(TASK_INSTRUCTION):
    print(str(ep))
    print()

What reflective_memory() recalls (with reflection):

  <episode>
    <task>Use the Monte Carlo method to estimate pi in one round.

IMPORTANT: MONTE CARLO ESTIMATES OF PI CONVERGE VERY SLOWLY. YOU WILL NEED MILLIONS OF POINTS FOR A RELIABLE RESULT.

Steps:
1. Call generate_random_sample(n) — choose n to maximise accuracy.
2. Call monte_carlo_estimate(sample_id) — this is your final estimate.

You have two tools:
- generate_random_sample(n): generates n random 2-D points in [0,1]x[0,1] and returns a sample_id.
- monte_carlo_estimate(sample_id): computes a pi estimate from the stored sample.

When done, respond with exactly:
{"sample_id": "<the-sample-id>"}

For this first attempt, use at most 1,000,000 points.</task>
    <result>The Monte Carlo estimate with 1,000,000 points is 3.139288.

{"sample_id": "8dbdb828-0e61-4280-81a8-002f4aad9e0a"}</result>
    <reflection>Use 10,000,000 points next time, because the Monte Carlo error scales roughly as \(1/\sqrt{n}\), so increasing the sample 

## Part 4 — A Recalled Lesson Guides the Next Run

The task is submitted a second time, this time without the one-million-point cap. Before the agent takes its first step,
`recall()` injects the episode from Part 2 — with its embedded reflection — into the
system prompt.

If Part 2 produced an inaccurate estimate, the reflection names exactly what to do
differently. If Part 2 succeeded, the reflection confirms the winning strategy.
Either way the agent starts this run with a concrete lesson rather than a blank
slate.

Check the logs: does the agent apply the recalled insight on its first tool call?

In [11]:
print("Memory injected into the system prompt for the second task:\n")
print(await memory.recall(Task(instruction=TASK_INSTRUCTION)))

Memory injected into the system prompt for the second task:

  <episode>
    <task>Use the Monte Carlo method to estimate pi in one round.

IMPORTANT: MONTE CARLO ESTIMATES OF PI CONVERGE VERY SLOWLY. YOU WILL NEED MILLIONS OF POINTS FOR A RELIABLE RESULT.

Steps:
1. Call generate_random_sample(n) — choose n to maximise accuracy.
2. Call monte_carlo_estimate(sample_id) — this is your final estimate.

You have two tools:
- generate_random_sample(n): generates n random 2-D points in [0,1]x[0,1] and returns a sample_id.
- monte_carlo_estimate(sample_id): computes a pi estimate from the stored sample.

When done, respond with exactly:
{"sample_id": "<the-sample-id>"}

For this first attempt, use at most 1,000,000 points.</task>
    <result>The Monte Carlo estimate with 1,000,000 points is 3.139288.

{"sample_id": "8dbdb828-0e61-4280-81a8-002f4aad9e0a"}</result>
    <reflection>Use 10,000,000 points next time, because the Monte Carlo error scales roughly as \(1/\sqrt{n}\), so increasing the

In [12]:
enable_console_logging(logging.INFO)

task4 = Task(instruction=TASK_INSTRUCTION)
result4 = await agent.run(task4)
print(result4)

INFO (llm_agents_fs.LLMAgent) :      🚀 Starting task: Use the Monte Carlo method to estimate pi in one round.

IMPORTANT: MONTE CARLO ESTIMATES OF PI CONVERGE VERY SLOWLY. YOU WILL NEED M...[TRUNCATED]
INFO (llm_agents_fs.TaskHandler) :      ⚙️ Processing Step: Use the Monte Carlo method to estimate pi in one round.

IMPORTANT: MONTE CARLO ESTIMATES OF PI CONVERGE VERY SLOWLY. YOU WILL NEE...[TRUNCATED]
INFO (llm_agents_fs.TaskHandler) :      🛠️ Executing Tool Call: generate_random_sample
INFO (llm_agents_fs.TaskHandler) :      ✅ Successful Tool Call: {"sample_id":"00215103-369b-4abe-8f1f-502008285694","sample_size":10000000}
INFO (llm_agents_fs.TaskHandler) :      ✅ Step Result: I need to make the following tool-calls:
{
    "id_": "1d65aff6-385c-48e2-85a2-925d25bafd26",
    "tool_name": "monte_carlo_estimate",
...[TRUNCATED]
INFO (llm_agents_fs.TaskHandler) :      🧠 New Step: Call the monte_carlo_estimate tool with sample_id "00215103-369b-4abe-8f1f-502008285694" to obtain the final 

## Key Takeaway

`reflective_memory()` converts task experience into transferable lessons. The
Reflexion paper frames this as verbal reinforcement learning: rather than updating
model weights, the agent stores natural-language feedback from each trial and
retrieves it when facing the same problem.

The Pi estimation task makes the mechanism concrete because the custom `template`
supplies the true value of pi, giving the reflection model an explicit error signal
to reason from. In the run above:

- **Task 1** was capped at 1 million points and produced an estimate of `3.139288`,
  an error of about 0.0023 from the true value.
- The reflection identified the 1/√N convergence rate and recommended **10 million
  points** by name as the next sample size.
- **Task 2** opened immediately with `generate_random_sample(10000000)`, applying the
  recalled lesson on its first tool call. It produced `3.1413264`, an error of about
  0.00027, roughly 9× closer. Ten times the points should cut the typical error by
  about 3×; this run did better than that partly by chance, since each run draws a
  single random sample.

The comparison in Parts 2 and 3 shows why the custom template matters. The
`recency_memory()` episode carries the same result (`3.139288`) but no lesson. A
future agent re-reading it would see the estimate, but not what sample size to try
next or why. The `reflective_memory()` episode distils the corrective lesson
in one sentence: a specific sample size, grounded in the mathematical reason it will
work better.

The factory takes three parameters: `llm` (used for reflection), `max_results` (how
many similar episodes to recall, defaults to `5`), and `template` (optional custom
reflection prompt, defaults to a general-purpose one-sentence lesson template).
Swapping in `recency_memory()` or `similarity_memory()` requires changing only the
one line that constructs the memory object.